server/src/main/java/org/apache/iotdb/db/query/fill/PreviousFill.java

```diff
@@ -59,28 +84,183 @@ public long getBeforeRange() {
     return beforeRange;
   }
 
+  @Override
+  public void configureFill(Path path, TSDataType dataType, long queryTime,
+      Set<String> sensors, QueryContext context)
+      throws StorageEngineException, QueryProcessException {
+    this.seriesPath = path;
+    this.dataType = dataType;
+    this.context = context;
+    this.queryTime = queryTime;
+    this.allSensors = sensors;
+    this.timeFilter = constructFilter();
+    this.dataSource = QueryResourceManager.getInstance().getQueryDataSource(path, context, timeFilter);
+    // update filter by TTL
+    timeFilter = dataSource.updateFilterUsingTTL(timeFilter);
+  }
+
   @Override
   public TimeValuePair getFillResult() throws IOException {
-    TimeValuePair beforePair = null;
-    TimeValuePair cachedPair;
-    while (batchData.hasCurrent() || allDataReader.hasNextBatch()) {
-      if (!batchData.hasCurrent() && allDataReader.hasNextBatch()) {
-        batchData = allDataReader.nextBatch();
+    TimeValuePair lastPointResult = retrieveValidLastPointFromSeqFiles();
+    UnpackOverlappedUnseqFiles(lastPointResult.getTimestamp());
+
+    long lastVersion = 0;
+    PriorityQueue<ChunkMetadata> sortedChunkMetatdataList = sortUnseqChunkMetadatasByEndtime();
+    while (!sortedChunkMetatdataList.isEmpty()
+        && lastPointResult.getTimestamp() <= sortedChunkMetatdataList.peek().getEndTime()) {
+      ChunkMetadata chunkMetadata = sortedChunkMetatdataList.poll();
+      TimeValuePair lastChunkPoint = getChunkLastPoint(chunkMetadata);
+      if (shouldUpdate(lastPointResult.getTimestamp(), lastVersion,
+          lastChunkPoint.getTimestamp(), chunkMetadata.getVersion())) {
+        lastPointResult = lastChunkPoint;
+        lastVersion = chunkMetadata.getVersion();
       }
-      cachedPair = new TimeValuePair(batchData.currentTime(), batchData.currentTsPrimitiveType());
-      batchData.next();
-      if (cachedPair.getTimestamp() <= queryTime) {
-        beforePair = cachedPair;
-      } else {
+    }
+    return lastPointResult;
+  }
+
+  /** Pick up and cache the last sequence TimeseriesMetadata that satisfies timeFilter */
+  private TimeValuePair retrieveValidLastPointFromSeqFiles() throws IOException {
+    List<TsFileResource> seqFileResource = dataSource.getSeqResources();
+    TimeValuePair lastPoint = new TimeValuePair(Long.MIN_VALUE, null);
+    for (int index = seqFileResource.size() - 1; index >= 0; index--) {
+      TsFileResource resource = seqFileResource.get(index);
+      TimeseriesMetadata timeseriesMetadata =
+          FileLoaderUtils.loadTimeSeriesMetadata(
+              resource, seriesPath, context, timeFilter, allSensors);
+      if (timeseriesMetadata != null) {
+        if (endtimeContainedByTimeFilter(timeseriesMetadata.getStatistics())) {
+          return constructLastPair(
+              timeseriesMetadata.getStatistics().getEndTime(),
+              timeseriesMetadata.getStatistics().getLastValue(),
+              dataType);
```

Comment:
Before you use the statistics info you should use the canUseStatistics() method in the Statistics class to judge whether the statistics can be used. Because may be the chunk has been modified.

Not only here, chunk statistics and page statistics are the same

Reply:
Fixed